packages/config/src/handlers/locations/locationTreeSolver.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * OpenCRVS is also distributed under the terms of the Civil Registration
+ * & Healthcare Disclaimer located at http://opencrvs.org/license.
+ *
+ * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
+ */
+
+import {
+  resourceIdentifierToUUID,
+  SavedLocation
+} from '@opencrvs/commons/types'
+import { UUID } from '@opencrvs/commons'
+
+/**
+ * Creates a new Map<SavedLocation.partOf, SavedLocation[]>
+ * It sets the first-level children under their parents
+ */
+const resolveParentChildrenMap = (locations: SavedLocation[]) => {
+  const parentChildrenMap = new Map<UUID, SavedLocation[]>()
+
+  for (const child of locations) {
+    if (!child.partOf) continue
+
+    const parentId = resourceIdentifierToUUID(child.partOf.reference)
+    const parentChildrenRelationship = parentChildrenMap.get(parentId)
+
+    if (!parentChildrenRelationship) {
+      parentChildrenMap.set(parentId, [child])
+    } else {
+      parentChildrenRelationship.push(child)
+    }
+  }
+
+  return parentChildrenMap
+}
+/** Resolves any given location's children multi-level down to the leaf node */
+export const resolveLocationChildren = (
+  parentId: UUID,
+  locations: SavedLocation[]
+) => {
+  const parentChildrenMap = resolveParentChildrenMap(locations)
+  const children: SavedLocation[] = []
+  const stack = parentChildrenMap.get(parentId) ?? []
+
+  while (stack.length) {
+    const child = stack.pop()!
+    children.push(child)
+    if (parentChildrenMap.get(child.id)) {
+      stack.push(...(parentChildrenMap.get(child.id) ?? []))
+    }
+  }
+
```

Comment:
Could this be done without while and immutable data structures? `while` always scares me a bit as it can get the whole process stuck in the worst case 😱

Reply:
I'll have a look if I can avoid the while-loop, with funky input parameters, it's possible to have an infinite loop here. In tests I was able to make it happen with impossible address structures. 

it's scary looking to me as well, but it's a common way to implement depth-first search and may be familiar to many devs with a glance